FP8 support on NVIDIA

pkotzbach opened the pull request:
without tensorcores, for simple_matmul.py i get
BFLOAT16:
`*** CUDA      21 r_128_64_8_16_1024_4_4_4                  arg  3 mem  0.13 GB tm     19.01ms/   117.17ms (  7229.39 GFLOPS    5.3|3616.5  GB/s) ['matmul']`

FP8E4M3:
`*** CUDA      24 r_128_64_8_16_1024_4_4_4                  arg  3 mem  0.07 GB tm     18.98ms/    98.23ms (  7241.72 GFLOPS    2.7|1811.3  GB/s) ['matmul']`
but without saturation (values are cut to highest possible value, no infinity)

with tencorcores:
BFLOAT16:
`*** CUDA      14 r_42_63_2_2_2_2_2_126_2_2_2_2_2_2_3_4     arg  3 mem  0.03 GB tm    364.22us/     3.15ms ( 44991.77 GFLOPS   67.0|2365.6  GB/s) ['matmul']`

FP8E4M3:
`*** CUDA      23 r_42_63_2_2_2_2_2_63_2_2_2_2_2_2_2_3_4    arg  3 mem  0.02 GB tm    306.53us/     4.71ms ( 53460.30 GFLOPS   39.8|1405.5  GB/s) ['matmul']`

continuation to #8490 (as i used wrong branch there)

ignaciosica commented:
Some high level comments:
- Resolve merge conflicts
- Many dtype tests failing in green box (don't trust CI as many tests are skipped)
- Python tensor core emulation is pending (ops_python). See #8597 for reference
- Check swizzling. Tensor cores tests are failing on green machine. I independently figured out the swizzle for `int8` tc (they share it with `f8`) and it has some discrepancies. 

---

![image](https://github.com/user-attachments/assets/528b2a77-2976-4ffb-b321-4ecd6d520622)

<img width="615" alt="image" src="https://github.com/user-attachments/assets/a33f277e-8fab-4dfe-9c7d-2088851f9174" />

ignaciosica commented:
As a good practice, keep it as draft until you are sure all tests pass. Pushing with broken tests is never a good idea. An mind that some tests are skipped in CI so you can't rely solely on it.

ignaciosica commented:
As a quick tip, you can verify the swizzle is correct independently from the precision error you get from the dtype itself, just by comparing the results you get with a matmul with TC=0. You should match that accuracy with TC=1. (There might be cases where this is not true, (TC not really accumulating on fp32) but I believe this is not the case).

ignaciosica commented on tinygrad/renderer/cstyle.py:
```diff
@@ -303,14 +304,23 @@ class CUDARenderer(CStyleLanguage):
   tc_8168_f16 = [TensorCore(dims=(8,16,8), threads=32, elements_per_thread=(4,2,4), dtype_in=dtypes.half, dtype_out=dtypes.float, opts=cuda_tc_opts,
     swizzle=(((6,7,2,3,4),(0,1,8,5,9)), ((6,7,8,0,1),(2,3,4,9,5))))]
+  tc_81632_f8 = [TensorCore(dims=(8,16,32), threads=32, elements_per_thread=(16,8,4), dtype_in=di, dtype_out=do, opts=cuda_tc_opts,
+    swizzle=(((7,8,2,3,4),(0,1,10,5,6,11,9)), ((10,0,1,7,8),(2,3,4,11,5,6,9))))
```
The swizzle is still slightly off. The only difference is that locals for B are out of order. First you move along the column and then you walk along the row, if that makes sense. I'll post the corrected swizzle


``` python
swizzle=(((7,8,2,3,4),(0,1,10,5,6,11,9)), ((7,8,10,0,1),(2,3,4,11,5,6,9))))
```

ignaciosica replied:
Threads are contiguous along the orange dotted line. See how first goes down along the column and then starts moving through the row.

<img src="https://github.com/user-attachments/assets/877e9995-81a4-48cd-913e-7320cacd14f2" width="256">

ignaciosica replied:
Before, threads where 'walking' along the row first and then going down through the column

<img src="https://github.com/user-attachments/assets/6d8aa452-fbfd-4d06-96e3-7b2e415c0336" width="256">

ignaciosica replied:
It was almost there though, good job!

pkotzbach commented:
i fixed the issues and i think its ready for review 
also (i dont know if its ok) i added dependency for fp8 numpy datatype, because struct.pack doesnt support fp8s, but its only for python cuda emulator

geohot commented on test/test_ops.py:
```diff
@@ -1133,6 +1133,18 @@ def test_small_gemm_eye(self):
   def test_gemm_fp16(self):
     helper_test_op([(64,64), (64,64)], lambda x,y: x.half().matmul(y.half()), atol=5e-3, rtol=5e-3)
+  @unittest.skipUnless("CUDA" in Device.get_available_devices(), "only if CUDA is available")
+  def test_gemm_fp8e4m3(self):
```
Don't copy paste code

geohot commented on tinygrad/renderer/cstyle.py:
```diff
@@ -337,7 +347,27 @@ def __reduce__(self): return self.__class__, (self.arch,)
-  type_map = {dtypes.bfloat16: "nv_bfloat16"}
+  type_map = {dtypes.bfloat16: "nv_bfloat16", dtypes.fp8e4m3: "__nv_fp8_e4m3", dtypes.fp8e5m2: "__nv_fp8_e5m2"}
+
+  extra_matcher = PatternMatcher([
```
Tons of copy paste!

geohot commented:
There's a lot of cleanup required here. Never copy paste code like this, think about the right abstraction

Also, I think the test failure might be real.

pkotzbach commented:
i did cleanup and fixed test

geohot commented on .github/workflows/test.yml:
```diff
@@ -477,7 +479,7 @@ jobs:
           cuda: ${{ (matrix.backend == 'ptx' || matrix.backend == 'triton' || matrix.backend == 'nv') && 'true' }}
       - name: Set env
-        run: printf "${{ matrix.backend == 'llvm' && 'LLVM=1' || matrix.backend == 'clang' && 'CLANG=1' || matrix.backend == 'gpu' && 'GPU=1' || matrix.backend == 'PTX' && 'FORWARD_ONLY=1\nJIT=1\nOPT=2\nCUDA=1\nPTX=1\nMOCKGPU=1' || matrix.backend == 'triton' && 'FORWARD_ONLY=1\nJIT=1\nOPT=2\nNV=1\nMOCKGPU=1\nTRITON=1\nTRITON_PTXAS_PATH=/usr/bin/ptxas' || matrix.backend == 'amd' && 'AMD=1\nMOCKGPU ...
+        run: printf "${{ matrix.backend == 'llvm' && 'LLVM=1' || matrix.backend == 'clang' && 'CLANG=1' || matrix.backend == 'gpu' && 'GPU=1' || matrix.backend == 'PTX' && 'FORWARD_ONLY=1\nJIT=1\nOPT=2\nCUDA=1\nPTX=1\nMOCKGPU=1' || matrix.backend == 'triton' && 'FORWARD_ONLY=1\nJIT=1\nOPT=2\nNV=1\nMOCKGPU=1\nTRITON=1\nTRITON_PTXAS_PATH=/usr/bin/ptxas' || matrix.backend == 'amd' && 'AMD=1\nMOCKGPU ...
```
What does this have to do with fp8?

geohot commented on extra/gemm/simple_matmul.py:
```diff
@@ -7,6 +7,25 @@
 if getenv("INT"):  dtype_in, acc_dtype = dtypes.int8, dtypes.int32
 if getenv("UINT"): dtype_in, acc_dtype = dtypes.uint8, dtypes.int32
 
+dtype_in = (
```
What is this spacing?

geohot commented on tinygrad/renderer/cstyle.py:
```diff
@@ -337,7 +344,17 @@ def __reduce__(self): return self.__class__, (self.arch,)
-  type_map = {dtypes.bfloat16: "nv_bfloat16"}
+  type_map = {dtypes.bfloat16: "nv_bfloat16", dtypes.fp8e4m3: "__nv_fp8_e4m3", dtypes.fp8e5m2: "__nv_fp8_e5m2"}
+  @staticmethod
+  def __create_fp8_patterns(dtype) -> list: return [(UPat(Ops.WHERE, src=(UPat.var("b"), UPat.var("x", dtype=dtype), UPat.var("y", dtype=dtype))),
```
??? formating

geohot commented on tinygrad/dtype.py:
```diff
@@ -158,8 +162,9 @@ def to_dtype(dtype:DTypeLike) -> DType: return dtype if isinstance(dtype, DType)
 # we don't support weak type and complex type
 promo_lattice = { dtypes.bool: [dtypes.int8, dtypes.uint8], dtypes.int8: [dtypes.int16], dtypes.int16: [dtypes.int32], dtypes.int32: [dtypes.int64],
-  dtypes.int64: [dtypes.float16, dtypes.bfloat16], dtypes.uint8: [dtypes.int16, dtypes.uint16], dtypes.uint16: [dtypes.int32, dtypes.uint32],
+  dtypes.int64: [dtypes.fp8e5m2, dtypes.fp8e4m3], dtypes.uint8: [dtypes.int16, dtypes.uint16], dtypes.uint16: [dtypes.int32, dtypes.uint32],
```
Does this match JAX?

pkotzbach replied:
according to promotion int->float it does, there is no explicit info about fp8s in JAX type promotion

geohot commented on setup.py:
```diff
@@ -65,7 +65,8 @@
             "ggml-python",
-            "capstone"
+            "capstone",
+            "ml_dtypes"
```
Interesting package. I'm fine with this.

geohot commented:
Getting close, a few spacing cleanups and anything I missed. Read the diff and polish it, then it should be good.

pkotzbach commented:
i think i've done my best with formatting

geohot commented:
Sorry this went stale, still interested in getting it merged?

chenyuxyz commented:
conflict, otherwise dtype stuff looks good. small issue that the least common type of int64 and uint64 becomes fp8 but not all backend supports it

chenyuxyz reviewed (commented):
overall looks good, left some minor testing comments. should be good to merge after

pkotzbach commented on test/test_dtype.py:
```diff
@@ -264,6 +284,7 @@ class TestBitCast(unittest.TestCase):
   def test_shape_change_bitcast(self, dt1, dt2):
     # NOTE: this has to be assume to prevent hypothesis from skipping all samples
     assume(dt2 != dtypes.bfloat16 and dt1 != dtypes.bfloat16) # no test for bf16 bitcast yet
+    assume(dt1 not in [dtypes.fp8e4m3, dtypes.fp8e5m2] and dt2 not in [dtypes.fp8e4m3, dtypes.fp8e5m2]) # no test for fp8 bitcast yet
```
yes it works with python backend but it doesnt with numpy and its used in test

pkotzbach replied:
it works internally on python backend, but when its casted to numpy backend it doesnt and its used in test
that would require changes in NpyAllocator to make it work

chenyuxyz commented:
getting 107 tflops `BEAM=4 FP8E4M3=1 DEBUG=2 python extra/gemm/simple_matmul.py` on 4090, the spec says 660. but somehow for bfloat16 i can only get 107.
<img width="1167" alt="image" src="https://github.com/user-attachments/assets/8c31e804-be58-494e-ab7c-fe18a3fc73aa" />

chenyuxyz commented:
Congrats! Bounty is yours, [george@tinygrad.org](mailto:george@tinygrad.org) to claim. Can pay via PayPal or USDC on ETH

geohot commented on tinygrad/runtime/ops_python.py:
```diff
@@ -100,11 +107,13 @@ def __call__(self, *bufs, global_size:tuple[int,int,int]=(1,1,1), local_size:tup
+          assert (dtp[0].fmt and dtype.fmt) or (dtp[0] in dtypes.fp8s and dtype) or (dtype in dtypes.fp8s and dtp[0])
+          if dtp[0] in dtypes.fp8s: packed = b''.join([truncate.get(dtp[0], lambda dt: dt)(z).tobytes() for z in inp[0]])
+          else: packed = struct.pack(str(warp_size) + str(dtp[0].fmt), *inp[0])
+          if dtype in dtypes.fp8s: ul[i] = np.frombuffer(packed,dtype=truncate.get(dtype, lambda x: x)).tolist()
```
Wrong!

geohot commented:
Several major bugs in this. @pkotzbach 

`DEBUG=2 EMULATE_CUDA_SM89=1 FORWARD_ONLY=1 PYTHON=1 python3 test/test_ops.py TestOpsFp8s.test_gemm_fp8e4m3` tests nothing! There's no CUDA device on the CI machine.

The Python device should not depend on numpy.

And this line `np.frombuffer(packed,dtype=truncate.get(dtype, lambda x: x)).tolist()` is complete nonsense, no idea how it got passed mypy (it didn't locally). You are passing a function in to a numpy dtype param

geohot commented:
That's in addition to bugs that were hotfixed.

https://github.com/tinygrad/tinygrad/commit/a3aaf92b210e89281c10969d9b9522f292707505 was opening all devices.
